### ISSUE TITLE
Fetching data for huge GitLAB repositories fails

### ISSUE THREAD

**Issue by dusek** [magit/forge#55]
```
Gitlab: error in process filter: ghub--signal-error: HTTP Error: 502, "Bad gateway"
```
## Steps to reproduce
1. `git clone git@gitlab.com/gitlab-org/gitlab-development-kit`
2. Open the repo from step 1. in magit
3. `F y`

## Expected results
Pull requests and issues would be fetched and displayed

## Actual results
After around 1 minute wait, this error message is displayed in `*Messages*` buffer (some of it flashes in minibuffer too):

```
error in process filter: ghub--signal-error: HTTP Error: 502, "Bad gateway", "/api/v4/projects/gitlab-org%2Fgitlab-development-kit/labels?per_page=100", ((message . "GitLab is not responding") (documentation_url . "https://github.com/magit/ghub/wiki/Github-Errors"))
error in process filter: HTTP Error: 502, "Bad gateway", "/api/v4/projects/gitlab-org%2Fgitlab-development-kit/labels?per_page=100", ((message . "GitLab is not responding") (documentation_url . "https://github.com/magit/ghub/wiki/Github-Errors"))
```

Magit status buffer's footer still shows `(Magit Pulling gitlab-org/gitlab-development-kit)` and keeps showing it forever (basically until I completely quit Emacs - now it has been already for more than 10 minutes).

Pull requests and issues are not displayed (and probably not fetched either).

## Configuration
- GNU Emacs 26.1 (build 1, x86_64-apple-darwin14.5.0, NS appkit-1348.17 Version 10.10.5 (Build 14F2511)) of 2018-05-31
- ghub               20181230.1822
- forge              20181230.1810
- magit              20181229.1434
- git version 2.20.1
- macOS 10.14.2 (18C54)

## Discussion
This is displayed in the `magit-process` buffer right after step 3.:

```
  0 git … config --add remote.origin.fetch \+refs/merge-requests/\*/head\:refs/pullreqs/\*
```

Originally experienced with the `gitlab-org/gitlab-ce` repo, but demoing this here on the `gitlab-development-kit` repo since it is magnitudes smaller than the former (thus enabling fast re-testing from scratch - clone takes ~5 seconds instead of ~5 minutes).

Doing Steps to reproduce for `git@github.com:magit/forge` yields Expected results (as a matter of fact, I am submitting this issue using `forge` from such cloned and setup repository right now).

**Comment by vkurup** [comment]
This may be a duplicate of #20. I switched to using the development version and haven't seen those 502 errors again.

**Comment by tarsius** [comment]
This isn't a duplicate.  What is common to both issues is that we get a 502 instead of something more useful and a difference is that in one case it is Github doing that and in the other Gitlab.

**Comment by tarsius** [comment]
I am afraid pulling data for such a humongous from Gitlab is not feasible.

Forge is designed around the idea that all data should be pulled and stored locally.  The fact that the initial pull can take a long time is a trade-off that we have to live with.  However at a certain size this becomes unusable and there isn't much we can do about it.  I was fully aware of that when I made this design decision, e.g. I decided to optimize for the best experience in 99% of all repositories at the cost of Forge being completely unusable in 1%.

I have just tried to pull this repository's API data myself and ran into the same error.

(At the time it happened Forge hadn't even progresses beyond fetching the list of forks.  That would be a problem even if no error occurred because that information is being fetched every time you pull, not just the first time.  Improvements could be made here, e.g. we could just fetch at most N forks (and labels and assignable users) and then stop.)

I also looked at the rate limiting headers and at least according to those, we weren't doing things to fast:

```
("RateLimit-Limit" . "600")
("RateLimit-Observed" . "1")
("RateLimit-Remaining" . "598")
```

Never-the-less I think we hit some limit.  It's not a limit that is observed by the part of Gitlab designed to detect that sort of thing though.  Instead it appears that some internal system gets overwhelmed resulting in something else concluding that "GitLab is not responding".

As mentioned before, we could just fetch only certain amount of X, or not fetch certain non-essential X at all, but even if we did that we might still run into some part of GitLab throwing up its arms with no way to recover.

With *Github* it is not quite as bad.  Pulling can still take very long (see #6), but at least eventually it finishes.  The fetching takes "only" tree hours in that case.  The fact that storing the fetched data also takes three hours is likely due to one or more bugs, which I expect can be fixed.  But here we don't even get to that point so it isn't really relevant.

With *Gitlab* there is no hope we will ever be able to fetch all the data for such huge repositories.  Like I proposed on #6 for *Github* we could let users specify to only fetch topics from the last N time units.  Additionally we could let users specify to not fetch things like forks *at all*.

That would make Forge usable for such huge *Gitlab* repositories.  I will implement that for *Github* first though.

However even once that is done fetching from *Gitlab* will still be *much* slower than for *Github*, the reason being that is *much* slower to fetch the same amount of data using Gitlab's REST API than it is to fetch the same using Github's GraphQL API.  There is not much we can do about that but to wait for Gitlab to [implement GraphQL support](https://docs.gitlab.com/ee/api/#road-to-graphql).

For small to medium sized repositories, this difference in speed doesn't matter much; so the initial fetch takes 10 minutes instead of half a minute.  Unfortunate but bearable.  But if it would take an hour to fetch a certain amount of data from Github, then fetching the same amount from Gitlab is simply not an option.  Especially because some error might occur after fetching for one day and you have to start all over.

----

So in summary, fetching all data for such a large repository from Gitlab is impossible.  Fetching just a subset is possible and I intend to implement support for that, but it will probably be a while until I get to that.

**Comment by dusek** [comment]
Thanks for taking the time to analyze that. I will keep my fingers crossed!

**Comment by tarsius** [comment]
I have pushed a quick draft to the `sparse-fetch`.  Please give that a try.  You will have to set both the lisp and the git variable.

I tried with "2018-12-25".  That took over 1.5 hours for 900 issues and 300 topics, most of which was spend on fetching topics one by one.  Unfortunately I forgot that I had to consider the value of `forge.omitExpensive` in more than one place, resulting in an error while storing.  Currently I am pulling with "2019-12-03" and it is still fetching the 226 issues.  It seems to me that using Forge on a huge repository hosted on Gitlab is not feasible even with this hack.  If you don't fetch for a few days, then even an update will take a long time.

**Comment by dusek** [comment]
I tried that with `(setq forge--initial-topic-until "2018-12-30")` and with `git config --global forge.omitExpensive true` on the git@gitlab.com:gitlab-org/gitlab-ce repo.

I have for 3 times in a row hit this error approx. after 5-10 minutes (always quit&relaunched Emacs between tries) in `*Messages*` buffer:

```
error in process sentinel: ghub--handle-response-headers: BUG: missing headers nil
error in process sentinel: BUG: missing headers nil
```

Sometime before that it is writing something like "Pulling issue 16/787" or "Pulling gitlab-org/gitlab-ce issues" in the buffer's "status footer" (though I have not witnessed what is written there *right before* the error message above appears, as I have to leave this running in the background and checking on it every few moments).

I will be happy to do any further debugging required.

**Comment by tarsius** [comment]
> I will be happy to do any further debugging required.

Thanks for the offer.  Prepare for a wild ride.

> (though I have not witnessed what is written there *right before* the error message above appears, as I have to leave this running in the background and checking on it every few moments).

Edit `forge--msg`:

``` diff
-    (when (and echo msg)
+    (when msg
```

> error in process sentinel: BUG: missing headers

I think the bug is not in Forge but either in `url` or on the server side.  I have been seeing this issue for a long time (thus the BUG message), but only very sporadically and never when I set out to debug it.  Since you can now reproduce it reliably I would be very thankful if you would narrow it down.

My guess is that either (1) the server sends an empty response, lacking even the headers, or that (2) `url` loses the response somewhere along the way, for unknown reasons.

One debug strategy could be to insert debug statements into `url.el` and related libraries, the goal being to figure out at what point the buffer that is supposed to contain the response becomes empty instead.  Since the buffer might even start out empty and because you have to wait 5-10 minutes whenever you adjust the debug statements, you should probably start by finding the place where you can check if the buffer starts out empty in the first place and add the initial set of debug statements there.

Before you do even that, try to reduce the data that has to be fetched before the issue is triggered (by increasing `forge--initial-topic-until`) to ensure you have to wait as little as possible before test runs.

**Comment by dusek** [comment]
I am sorry but I will unfortunately not be able to be debugging this any time soon, unlike I originally indicated. Letting know here also so that anyone else who is hitting this as well feels free to give it a try.

I will just add that when I set the `forge--initial-topic-until` to the same day as the day I was trying it on, I was able to at least once successfully fetch the issues (35) and pull requests (16). So one might perhaps have better luck reproducing this with higher volume of days (e.g. 2 days in range instead of just 1).

I also wanted to try mitmproxy as first stage, but unfortunately hit mitmproxy/mitmproxy#848 (I hit that btw. also with just `list-packages`).

**Comment by tarsius** [comment]
> So in summary, fetching all data for such a large repository from Gitlab is impossible. Fetching just a subset is possible and I intend to implement support for that, but it will probably be a while until I get to that.

Done that now. Invoke `forge-pull` with a prefix argument (and also see its doc-string).
